I create the below GitHub action and test it:
```
name: Test Setup-Hooks Script

on:
  workflow_dispatch:

jobs:
  test-script:
    runs-on: windows-latest

    steps:
      # Checkout repository
      - name: Checkout code
        uses: actions/checkout@v4

      # Install pre-commit
      - name: Install pre-commit
        run: |
          pip install pre-commit
          pre-commit --version

      # Install Node.js (for npm)
      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 18
      - run: npm --version

      # Ensure Git is installed
      - name: Check Git version
        run: git --version

      # Run the setup-hooks script
      - name: Execute setup-hooks script
        run: |
          pwsh ./scripts/setup-hooks.ps1

      - name: Invalid Commit
        run: |
          echo "sample content" > file.txt
          git add .
          git commit -m "Initial commit"
```

I got the output below, so everything seems fine.

```txt
Starting setup-hooks script...
Installing 'commit-msg' pre-commit hook...
pre-commit installed at .git\hooks\commit-msg
Installing 'prepare-commit-msg' pre-commit hook...
pre-commit installed at .git\hooks\prepare-commit-msg
Installing 'pre-merge-commit' pre-commit hook...
pre-commit installed at .git\hooks\pre-merge-commit
Installing 'pre-push' pre-commit hook...
pre-commit installed at .git\hooks\pre-push
Enabling 'pre-rebase' hook...
Template hook copied to D:\a\test-powershell\test-powershell\.git\hooks\pre-rebase.
Installing npm dependencies...

added 101 packages, and audited 102 packages in 7s

19 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Setup-hooks script completed successfully.
```

```
don't commit to branch...................................................Failed
- hook id: no-commit-to-branch
- exit code: 1

Error: Process completed with exit code 1.
```

--- a/scripts/setup-hooks.ps1
+++ b/scripts/setup-hooks.ps1
@@ -1,98 +1,128 @@
 #!/usr/bin/env pwsh
-# Usage:
-#   .\scripts\setup-hooks.ps1 (execute script)
 
-# Function to enable a specific Git hook by copying its sample file
+<#
+.SYNOPSIS
+    Setup Git hooks and npm dependencies for the repository.
+
+.DESCRIPTION
+    This script installs pre-commit hooks, enables additional Git hooks, 
+    and optionally installs npm dependencies if package.json exists.
+
+    Usage:
+        ./setup-hooks.ps1
+#>
+
+# Enable strict mode to catch errors and undefined variables
+Set-StrictMode -Version Latest
+
+# Enable a specific Git hook by renaming its sample file
 function Enable-Hook {
     param (
-        [Parameter(Mandatory)][string]$HooksDir,
-        [Parameter(Mandatory)][string]$HookName
+        [string]$HooksDir,
+        [string]$HookName
     )
 
-    $TargetHook = Join-Path $HooksDir $HookName
-    $TemplateFile = Join-Path $HooksDir "$HookName.sample"
+    $TargetHook = Join-Path -Path $HooksDir -ChildPath $HookName
+    $TemplateFile = Join-Path -Path $HooksDir -ChildPath "$HookName.sample"
 
     if (Test-Path $TargetHook) {
-        Write-Host "The '$HookName' hook is already enabled. ($TargetHook)" -ForegroundColor Yellow
+        Write-Host "[INFO] $HookName hook is already enabled. ($TargetHook)"
     } elseif (Test-Path $TemplateFile) {
-        Write-Host "Enabling '$HookName' hook..." -ForegroundColor Green
+        Write-Host "[INFO] Enabling $HookName hook..."
         Move-Item -Path $TemplateFile -Destination $TargetHook -Force
-        Write-Host "Template hook copied to $TargetHook." -ForegroundColor Green
+        Write-Host "[SUCCESS] $HookName hook enabled at $TargetHook."
     } else {
-        Write-Warning "Template for '$HookName' hook not found."
+        Write-Host "[WARNING] Template for $HookName hook not found. Skipping..."
     }
 }
 
-# Function to get the Git hooks path
+# Get the root directory of the Git repository
+function Get-RootPath {
+    $RootPath = git rev-parse --show-toplevel 2>$null
+    if (-not $RootPath) {
+        Write-Host "[ERROR] Not a Git repository."
+        exit 1
+    }
+    return $RootPath
+}
+
+# Get the path to package.json, if it exists
+function Get-PackageJsonPath {
+    $RootPath = Get-RootPath
+    $PackageJsonPath = Join-Path -Path $RootPath -ChildPath "package.json"
+    if (Test-Path $PackageJsonPath) {
+        return $PackageJsonPath
+    } else {
+        return $null
+    }
+}
+
+# Get the Git hooks directory path
 function Get-HooksPath {
-    try {
-        $GitRoot = git rev-parse --show-toplevel 2>$null
-        $HooksPath = git rev-parse --git-path hooks 2>$null
-
-        if ($GitRoot -and $HooksPath) {
-            return Join-Path $GitRoot $HooksPath
-        }
-    } catch {
-        Write-Warning "Unable to determine Git hooks path. Ensure this script is run inside a Git repository."
+    $HooksPath = git rev-parse --git-path hooks 2>$null
+    if (-not $HooksPath) {
+        Write-Host "[ERROR] Failed to locate Git hooks directory."
+        exit 1
     }
-
-    return ""
+    return $HooksPath
 }
 
-# Function to check if a command exists
+# Check if a command exists in the system
 function Command-Exists {
-    param ([Parameter(Mandatory)][string]$Command)
+    param (
+        [string]$Command
+    )
 
-    return $null -ne (Get-Command $Command -ErrorAction SilentlyContinue)
+    Get-Command $Command -ErrorAction SilentlyContinue | Out-Null
+    return $?
 }
 
-# Function to install pre-commit hooks
-function Install-PreCommitHooks {
-    param ([Parameter(Mandatory)][string[]]$HookTypes)
+# Main function
+function Main {
+    # Ensure `pre-commit` is installed
+    if (-not (Command-Exists "pre-commit")) {
+        Write-Host "[ERROR] pre-commit is not installed. Please install it first: https://pre-commit.com/#install"
+        exit 1
+    }
 
-    foreach ($HookType in $HookTypes) {
-        Write-Host "Installing '$HookType' pre-commit hook..." -ForegroundColor Cyan
+    # Check for package.json and npm
+    $PackageJsonPath = Get-PackageJsonPath
+    if ($PackageJsonPath -and -not (Command-Exists "npm")) {
+        Write-Host "[ERROR] npm is required to install dependencies but is not installed."
+        exit 1
+    }
+
+    # Install pre-commit hooks
+    Write-Host "============================"
+    Write-Host "Installing pre-commit hooks"
+    Write-Host "============================"
+    $PreCommitHooks = @("commit-msg", "prepare-commit-msg", "pre-merge-commit", "pre-push")
+    foreach ($HookType in $PreCommitHooks) {
+        Write-Host "[INFO] Installing $HookType hook..."
         pre-commit install --hook-type $HookType
+        Write-Host "[SUCCESS] $HookType hook installed."
     }
-}
 
-# Function to install npm dependencies
-function Install-NpmDependencies {
-    if (Test-Path "package.json") {
-        Write-Host "Installing npm dependencies..." -ForegroundColor Cyan
-        npm install
-    } else {
-        Write-Host "No package.json found. Skipping npm dependency installation." -ForegroundColor Yellow
-    }
-}
-
-# Main Script Execution
-Write-Host "Starting setup-hooks script..." -ForegroundColor Cyan
-
-# Check for required commands
-if (-not (Command-Exists "pre-commit")) {
-    Write-Error "pre-commit is not installed. Please install it first."
-    exit 1
-}
-
-if ((-not (Command-Exists "npm")) -and (Test-Path "package.json")) {
-    Write-Error "npm is not installed. Please install it first."
-    exit 1
-}
-
-# Install pre-commit hooks
-$PreCommitHooks = @("commit-msg", "prepare-commit-msg", "pre-merge-commit", "pre-push")
-Install-PreCommitHooks -HookTypes $PreCommitHooks
-
-# Enable additional Git hooks if hook path exists
-$HooksDir = Get-HooksPath
-if ($HooksDir) {
+    # Enable additional Git hooks
+    Write-Host "============================"
+    Write-Host "Enabling additional Git hooks"
+    Write-Host "============================"
+    $HooksDir = Get-HooksPath
     Enable-Hook -HooksDir $HooksDir -HookName "pre-rebase"
-} else {
-    Write-Warning "Git hooks directory not found. Skipping additional hook setup."
+
+    # Install npm dependencies if package.json exists
+    if ($PackageJsonPath) {
+        Write-Host "============================"
+        Write-Host "Installing npm dependencies"
+        Write-Host "============================"
+        npm install
+        Write-Host "[SUCCESS] npm dependencies installed."
+    }
+
+    Write-Host "============================"
+    Write-Host "[SUCCESS] Setup completed."
+    Write-Host "============================"
 }
 
-# Install npm dependencies
-Install-NpmDependencies
-
-Write-Host "Setup-hooks script completed successfully." -ForegroundColor Green
+# Execute the main function
+Main

--- a/scripts/setup-hooks.ps1
+++ b/scripts/setup-hooks.ps1
@@ -1,0 +1,98 @@
+#!/usr/bin/env pwsh
+# Usage:
+#   .\scripts\setup-hooks.ps1 (execute script)
+
+# Function to enable a specific Git hook by copying its sample file
+function Enable-Hook {
+    param (
+        [Parameter(Mandatory)][string]$HooksDir,
+        [Parameter(Mandatory)][string]$HookName
+    )
+
+    $TargetHook = Join-Path $HooksDir $HookName
+    $TemplateFile = Join-Path $HooksDir "$HookName.sample"
+
+    if (Test-Path $TargetHook) {
+        Write-Host "The '$HookName' hook is already enabled. ($TargetHook)" -ForegroundColor Yellow
+    } elseif (Test-Path $TemplateFile) {
+        Write-Host "Enabling '$HookName' hook..." -ForegroundColor Green
+        Move-Item -Path $TemplateFile -Destination $TargetHook -Force
+        Write-Host "Template hook copied to $TargetHook." -ForegroundColor Green
+    } else {
+        Write-Warning "Template for '$HookName' hook not found."
+    }
+}
+
+# Function to get the Git hooks path
+function Get-HooksPath {
+    try {
+        $GitRoot = git rev-parse --show-toplevel 2>$null
+        $HooksPath = git rev-parse --git-path hooks 2>$null
+
+        if ($GitRoot -and $HooksPath) {
+            return Join-Path $GitRoot $HooksPath
+        }
+    } catch {
+        Write-Warning "Unable to determine Git hooks path. Ensure this script is run inside a Git repository."
+    }
+
+    return ""
+}
+
+# Function to check if a command exists
+function Command-Exists {
+    param ([Parameter(Mandatory)][string]$Command)
+
+    return $null -ne (Get-Command $Command -ErrorAction SilentlyContinue)
+}
+
+# Function to install pre-commit hooks
+function Install-PreCommitHooks {
+    param ([Parameter(Mandatory)][string[]]$HookTypes)
+
+    foreach ($HookType in $HookTypes) {
+        Write-Host "Installing '$HookType' pre-commit hook..." -ForegroundColor Cyan
+        pre-commit install --hook-type $HookType
+    }
+}
+
+# Function to install npm dependencies
+function Install-NpmDependencies {
+    if (Test-Path "package.json") {
+        Write-Host "Installing npm dependencies..." -ForegroundColor Cyan
+        npm install
+    } else {
+        Write-Host "No package.json found. Skipping npm dependency installation." -ForegroundColor Yellow
+    }
+}
+
+# Main Script Execution
+Write-Host "Starting setup-hooks script..." -ForegroundColor Cyan
+
+# Check for required commands
+if (-not (Command-Exists "pre-commit")) {
+    Write-Error "pre-commit is not installed. Please install it first."
+    exit 1
+}
+
+if ((-not (Command-Exists "npm")) -and (Test-Path "package.json")) {
+    Write-Error "npm is not installed. Please install it first."
+    exit 1
+}
+
+# Install pre-commit hooks
+$PreCommitHooks = @("commit-msg", "prepare-commit-msg", "pre-merge-commit", "pre-push")
+Install-PreCommitHooks -HookTypes $PreCommitHooks
+
+# Enable additional Git hooks if hook path exists
+$HooksDir = Get-HooksPath
+if ($HooksDir) {
+    Enable-Hook -HooksDir $HooksDir -HookName "pre-rebase"
+} else {
+    Write-Warning "Git hooks directory not found. Skipping additional hook setup."
+}
+
+# Install npm dependencies
+Install-NpmDependencies
+
+Write-Host "Setup-hooks script completed successfully." -ForegroundColor Green

